src/Microsoft.Framework.CopyOnWriteDictionary.Internal/CopyOnWriteDictionary.cs

```diff
@@ -0,0 +1,148 @@
+﻿// Copyright (c) Microsoft Open Technologies, Inc. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections;
+using System.Collections.Generic;
+
+namespace Microsoft.Framework.Internal
+{
+    /// <summary>
+    /// A <see cref="IDictionary{TKey, TValue}"/> that defers creating a shallow copy of the source dictionary until
+    ///  a mutative operation has been performed on it.
+    /// </summary>
+    public class CopyOnWriteDictionary<TKey, TValue> : IDictionary<TKey, TValue>
+    {
+        private readonly IDictionary<TKey, TValue> _sourceDictionary;
+        private readonly IEqualityComparer<TKey> _comparer;
+        private IDictionary<TKey, TValue> _innerDictionary;
+
+        public CopyOnWriteDictionary([NotNull] IDictionary<TKey, TValue> sourceDictionary,
+                                     [NotNull] IEqualityComparer<TKey> comparer)
+        {
+            _sourceDictionary = sourceDictionary;
+            _comparer = comparer;
+        }
+
+        private IDictionary<TKey, TValue> ReadDictionary
+        {
+            get
+            {
+                return _innerDictionary ?? _sourceDictionary;
+            }
+        }
+
+        private IDictionary<TKey, TValue> WriteDictionary
+        {
+            get
+            {
+                if (_innerDictionary == null)
+                {
+                    _innerDictionary = new Dictionary<TKey, TValue>(_sourceDictionary,
+                                                                    _comparer);
+                }
+
+                return _innerDictionary;
+            }
+        }
+
+        public virtual ICollection<TKey> Keys
```

Comment:
This type is internal, don't need any doc comments at all. If we happen to writ something useful, that's great, but we definitely don't to go out of our way here. And tpyos and whatnot don't matter here either, beyond basic professionalism.